skyrim/common.py

```diff
@@ -0,0 +1,157 @@
+import hashlib
+import base58
+import time
+import xarray as xr
+import boto3
+import s3fs
+import os
+from datetime import datetime
+from typing import Literal
+from pathlib import Path
+from loguru import logger
+from urllib.parse import urlparse
+from io import BytesIO
+
+
+def generate_forecast_id(length=10):
+    """
+    Generate a unique forecast ID using the current time.
+    """
+    current_time = str(time.time()).encode()
+    hash_object = hashlib.sha256(current_time)
+    digest = hash_object.digest()
+    base58_encoded = base58.b58encode(digest)
+    return base58_encoded.decode()[:length]
+
+
+def generate_filename(
+    model: str,
+    start_time: datetime,
+    pred_time: datetime,
+    ic_source: Literal["cds", "file", "ifs", "gfs"] = "cds",
+):
+    """
+    Generate a filename for the forecast file.
+    args:
+    model: str: model name
+    start_time: datetime: start time of the forecast
+    pred_time: datetime: prediction time of the forecast
+    ic_source: Literal["cds", "file", "ifs", "gfs"]: initial condition source
+    """
+    return (
+        f"{model}"
+        + "__"
+        + f"{ic_source}__"
+        + f"{start_time.strftime('%Y%m%d_%H:%M')}"
+        + "__"
+        + f"{pred_time.strftime('%Y%m%d_%H:%M')}.nc"
+    )
+
+
+OUTPUT_DIR = str(Path.cwd() / "outputs")
+
+if not Path(OUTPUT_DIR).exists():
+    Path(OUTPUT_DIR).mkdir()
+    logger.success(f"Created output directory: {OUTPUT_DIR}")
+
+DEFAULT_SAVE_CONFIG = {
+    "output_dir": OUTPUT_DIR,
+    "file_type": "netcdf",
+    "filter_vars": [],
+}
+
+
+def remote_forecast_exists(path: str):
+    if not path.endswith("/"):
+        path += "/"
+    s3_client = boto3.client("s3")
+    p = urlparse(path)
+    return "Contents" in s3_client.list_objects_v2(
+        Bucket=p.netloc, Prefix=p.path[1:], MaxKeys=1
+    )
+
+
+def save_forecast(
+    pred: xr.DataArray | xr.Dataset,
```

Comment:
Currently, all model predictions are returned as `xr.DataArray`  types except the graphcast afair. We better homogenize the data types in the near future ⏳

Reply:
agree, also this can be used within Prediction so that we do prediction.save(...)